src/main/java/org/opensearch/knn/index/codec/KNN9120Codec/DerivedSourceStoredFieldsWriter.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.KNN9120Codec;
+
+import lombok.RequiredArgsConstructor;
+import org.apache.lucene.codecs.StoredFieldsWriter;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.MergeState;
+import org.apache.lucene.store.DataInput;
+import org.apache.lucene.util.BytesRef;
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.common.io.stream.BytesStreamOutput;
+import org.opensearch.common.xcontent.XContentHelper;
+import org.opensearch.common.xcontent.support.XContentMapValues;
+import org.opensearch.core.common.bytes.BytesReference;
+import org.opensearch.core.xcontent.MediaType;
+import org.opensearch.core.xcontent.MediaTypeRegistry;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.mapper.SourceFieldMapper;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+@RequiredArgsConstructor
+public class DerivedSourceStoredFieldsWriter extends StoredFieldsWriter {
+
+    private final StoredFieldsWriter delegate;
+    private final List<String> vectorFieldTypes;
+
+    @Override
+    public void startDocument() throws IOException {
+        delegate.startDocument();
+    }
+
+    @Override
+    public void writeField(FieldInfo fieldInfo, int i) throws IOException {
+        delegate.writeField(fieldInfo, i);
+    }
+
+    @Override
+    public void writeField(FieldInfo fieldInfo, long l) throws IOException {
+        delegate.writeField(fieldInfo, l);
+    }
+
+    @Override
+    public void writeField(FieldInfo fieldInfo, float v) throws IOException {
+        delegate.writeField(fieldInfo, v);
+    }
+
+    @Override
+    public void writeField(FieldInfo fieldInfo, double v) throws IOException {
+        delegate.writeField(fieldInfo, v);
+    }
+
+    @Override
+    public void writeField(FieldInfo info, DataInput value, int length) throws IOException {
+        delegate.writeField(info, value, length);
+    }
+
+    @Override
+    public int merge(MergeState mergeState) throws IOException {
+        return delegate.merge(mergeState);
+    }
+
+    @Override
+    public void writeField(FieldInfo fieldInfo, BytesRef bytesRef) throws IOException {
+        // Parse out the vectors from the source
+        if (Objects.equals(fieldInfo.name, SourceFieldMapper.NAME) && !vectorFieldTypes.isEmpty()) {
+            Tuple<? extends MediaType, Map<String, Object>> mapTuple = XContentHelper.convertToMap(
```

Comment:
lets have some timing logs here. I think this is going to be an expensive operation for converting bytes to JSON

Reply:
Will add in another review